---
title: "Demonstrating the Power of Non-Euclidean Embedding with Simulations"
author: "Russell Dinnage"
date: "26/06/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
knitr::knit_hooks$set(webgl = rgl::hook_webgl)
```

## Non-Euclidean Embeddings

Genetic data is high-dimensional. Hundreds of thousands to millions of nucleotides or SNPs for every individual. Because of this biologists have turned to dimension reduction techniques for visualizing and exploring genetic datasets, and to attempt to develop intuitions for it. This means that dimension reduction methods are extremely important in the biologist's toolkit. Despite this, the majority of biologists do not extend much beyond a simple PCA for initialy visualisation, and in some case, for important downstream analyses. This is a shame because while PCA is a useful technique that has its place in our toolbox, there is now a plethora of sophisticated methods for this task that should be greater explored in the field of molecular biology, evolution and ecology. In particular, the burgeoning computationally field of manifold learning is ripe to be better exploited in biology. 

This vignette only scratches the surface of this by showing how a simple choice between two kinds of lower dimensional embeddings can have an important effect on interpretation. 

This is based on recipe 10.4.1 from the SLiM manual

```
initialize() {
  initializeMutationRate(1e-7);
  initializeMutationType("m1", 0.5, "f", 0.0); // neutral
  initializeMutationType("m2", 0.5, "f", 0.1); // balanced
  initializeGenomicElementType("g1", c(m1,m2), c(999,1));
  initializeGenomicElement(g1, 0, 99999);
  initializeRecombinationRate(1e-8);
}

1 { sim.addSubpop("p1", 500); }

10000 { sim.simulationFinished(); }

fitness(m2) {
  return 1.5 - sim.mutationFrequencies(p1, mut);
}
```

```{r specify_model}
library(slimr)
set.seed(12345)

slim_script(
  
  slim_block(initialize(), {
    initializeMutationRate(1e-5);
    initializeMutationType("m1", 0.5, "f", 0.0); ## neutral
    initializeMutationType("m2", 0.5, "f", 0.1); ## balanced
    initializeMutationType("m3", 0.5, "f", 0.1); ## balanced
    initializeMutationType("m4", 0.5, "f", 0.1); ## balanced
    initializeGenomicElementType("g1", m1, 1);
    initializeGenomicElementType("g2", c(m1, m2), c(0.1, 0.9));
    initializeGenomicElementType("g3", c(m1, m3), c(0.1, 0.9));
    initializeGenomicElementType("g4", c(m1, m4), c(0.1, 0.9));
    ## place 3 blocks of each selected mutation type in each third of the genome
    initializeGenomicElement(c(g1, 
                               g2,
                               g1,
                               g2,
                               g1,
                               g2,
                               g1,
                               g3,
                               g1,
                               g3,
                               g1,
                               g3, 
                               g1,
                               g4,
                               g1,
                               g4,
                               g1,
                               g4,
                               g1), 
                             start = c(0, 
                                       1111, 
                                       1312, 
                                       2222, 
                                       2423,
                                       3333,
                                       3534,
                                       4444,
                                       4645,
                                       5555,
                                       5756,
                                       6666,
                                       6867,
                                       7777,
                                       7978,
                                       8888,
                                       9089,
                                       9999,
                                       12000), 
                             end =   c(1110, 
                                       1311, 
                                       2221, 
                                       2422,
                                       3332,
                                       3533,
                                       4443,
                                       4644,
                                       5554,
                                       5755,
                                       6665,
                                       6866,
                                       7776,
                                       7977,
                                       8887,
                                       9088,
                                       9998,
                                       11099,
                                       13109));
    initializeRecombinationRate(1e-8);
  }),
  
  slim_block(1, {
    sim.addSubpop("p1", 500);
  }),
  
  slim_block(1, 5000, late(), {
    slimr_output( sim.outputFull() , "fix_mut", do_every = 10);
  }),
  
  slim_block(5000, {
    sim.simulationFinished();
  }),
  
  slim_block(fitness(m2, p1), {
    dominance = asInteger(homozygous) * 0.5 + 0.5;
    return(1.5 - sim.mutationFrequencies(subpop, mut) * dominance);
  })
  
) -> script_1

script_1
```

For the purposes of testing, we've added stabilizing selection to one of the mutation types (m2). Later in this vignette we will use some of the advanced features of `slimr` to add custom stabilizing selection to each mutation type which is subpopulation specific. But first let's check this script works.

```{r test_script}
test <- slim_run(script_1)
test_muts<- slim_outputFull_extract(test$output_data, "mutations")
test_muts

```

```{r plot_equi}
library(ggplot2)
m2 <- test_muts %>%
  dplyr::filter(mut_type == "m2")

ggplot2::ggplot(m2, ggplot2::aes(generation, prevalence)) +
  ggplot2::geom_path(ggplot2::aes(group = as.factor(unique_id), 
                                  colour = as.factor(unique_id)), alpha = 0.25) +
  ggplot2::scale_y_continuous(limits = c(0, 1000)) +
  ggplot2::theme_minimal() +
  theme(legend.position = "none")
```

So we can see that the m2 mutations each tend to rise to and are kept at an intermediate frequency by stabilizing selection. It stays at a frequency of roughly 0.5 (bearing in mind there are 500 diploid individuals, leading to 1000 genomes in the population). This equilibrium value is the result of the simple fitness callback we used, where a stabilizing relative fitness of 1.0 is reached at 0.5 (1.5 - 0.5). If frequency is higher than 0.5, the relative fitness become less than 1.0, if frequency is lower, it becomes greater than 1.0, so the mutation will tend to fluctuate around this value. We can create a general equation to produce stabilizing selection like this `rel_fitness = (1 + eq) - mut_freq`, where `eq` is the desired equilibrium frequency, and `mut_freq` is the current frequency of the mutation.

Okay, so why did we set up this model? We can use this model to simulate subpopulation with different pattern of selection on a small number of genomic sites. We can use this set up then to generate subpopulation that differ on a known number of dimensions (3 in this case, with three types of mutations). We can then use dimension reduction techniques to see if they can accurately represent patterns we generate in the low-dimensional space, in the presence of loads of noise from the background evolutionary processes being simulated by SLiM, such as drift, background neutral mutations, and linkage disequilibrium with the sites under selection. As an additional wrinkle, we will generate patterns in only two effective dimensions. It will become clear as we go along, why this is a useful thing to do.

## Create low-dimensional patterns - comparing planar and sperical geometry

We will generate some random data to represent a set of selection coefficients for a number of subpopulations. We will simulate 40 subpopulations, initially. We will first generate random points in 3d space, which we will then thin out such that most of the information is contained in only two dimensions. We will use two strategies for this, 1) thinning to a 2d plane, 2) thinning to the surface of a 3d sphere. These two scenarios will both effectively only make use of 2 dimensions, but the difference will be apparent in our dimension reduction. The following could be done in multiple ways, but to further showcase the power of combining SLiM simulation and the vast array of powerful R packages, we will use two packages designed for manipulating 3d objects in R for this next part, `Rvcg` and `Morpho`. As a bonus, we can easily visualise our results in 3d using the `rgl` package. First we generate a 3d point cloud, then create objects representing planes, and spheres.

```{r make_random_sel}
library(Rvcg)
library(rgl)
library(proxy)

pnts <- cbind(runif(100000, -0.5, 0.5), 
              runif(100000, -0.5, 0.5), 
              runif(100000, -0.5, 0.5))

rgl::points3d(pnts)

```

Now  we create a random plane, calculate the distance of our points to it, then filter them to only those very close to the plane (less than 0.01 distance).

```{r random_plane, webgl=TRUE, results='hide'}
random_plane <- list(V1 = rnorm(3), v2 = rnorm(3), v3 = rnorm(3))
pnts_on_plane <- Morpho::points2plane(pnts, v1 = random_plane$V1,
                                      v2 = random_plane$v2,
                                      v3 = random_plane$v3)
pnt_dists <- proxy::dist(pnts, pnts_on_plane, pairwise = TRUE)
plane_pts <- pnts[pnt_dists < 0.01, ][1:40, ]

rgl::open3d()
rgl::spheres3d(plane_pts, radius = 0.05)

```

We now repeat a similar procedure for the spherical points. We calculate the distance between our random points to a sphere (of radius 2), then only retain those very close to it. Generally 

```{r sphere_pnts, webgl=TRUE, results='hide'}
sphere <- Rvcg::vcgSphere() %>%
  Morpho::scalemesh(0.5)
pnts_on_sphere <- Rvcg::vcgClost(pnts, sphere)
sphere_pts <-  Morpho::vert2points(pnts_on_sphere)[abs(pnts_on_sphere$quality) < 0.01, ]
sphere_pts <- sphere_pts[sample.int(nrow(sphere_pts), 40), ]

rgl::open3d()
rgl::spheres3d(sphere_pts, radius = 0.05)
```
Now that we have some 3d points, we can translate these into equilibrium frequencies for our set of three mutations types. We just need to scale them between 0 and 1 as a start.

```{r scale_pts}
plane_pts <- plane_pts + 0.5
sphere_pts <- sphere_pts + 0.5

sphere_pts
```
As a sanity check we can make sure out spherical points have nearly the same Euclidean norm (from 0.5), equal or close to 0.5.
```{r sanity_check}
apply(sphere_pts, 1, function(x) sqrt(sum((x-0.5)^2)))
```

We first need to make a new `slimr_script`, where we can include subpopulation specific equilibrium frequencies for our mutations. We will do this using some advanced features of `slimr`, specifically, by using 'non-standard evaluation' or NSE, as exemplified by the `rlang` package. We will make use of the 'forcing' operator `!!`, as well as the 'force-splice' operator `!!!`. `slimr` fully supports these operators, which do what they sound like, they force part of R code to be evaluated selectively, and their results inserted directly into code. This 'metaprogramming' technique allows us to do some powerful things in conjunction with SLiM through `slimr`. Here we will programmatically generate SLiM code blocks to specify custom `fitness()` callback, giving our mutations their desired equilibrium frequencies. First let's create these code blocks, which we can then 'splice' into a modified `slim_script` call. We will use a set of loops to make it obvious what we are doing here, but you could also use `purrr` for this purpose. SLiM code passed into `slim_block` is usually interpreted as verbatim code, with the exception of special `slimr` function such as `slimr_template`, `slimr_output`, etc.. This means that if you want to pass in the result of an R function, you need to use a special operator to let slimr know that the code is not meant to be SLiM code, but rather you want it evaluated in R before becoming SLiM code. Here is an example where the result may not be what you are expecting:

```{r example_mistake}
x <- 1
slim_script(
  slim_block(initialize(), {
    initializeMutationRate(1e-7); 
    initializeMutationType("m1", 0.5, "f", 0.0);
    initializeGenomicElementType("g1", m1, 1.0);
    initializeGenomicElement(g1, 0, 99999);
    initializeRecombinationRate(1e-8);
  }),
  slim_block(1, {
    sim.addSubpop(paste("p", x, sep = ""), 500)    
  })
) -> mistake

mistake

slim_run(mistake, simple_run = TRUE, show_output = TRUE)
```

```{r example_mistake_fix}
x <- 1
slim_script(
  slim_block(initialize(), {
    initializeMutationRate(1e-7); 
    initializeMutationType("m1", 0.5, "f", 0.0);
    initializeGenomicElementType("g1", m1, 1.0);
    initializeGenomicElement(g1, 0, 99999);
    initializeRecombinationRate(1e-8);
  }),
  slim_block(1, {
    sim.addSubpop(!!paste("p", x, sep = ""), 500)    
  })
) -> mistake_fixed

mistake_fixed

slim_run(mistake_fixed, simple_run = TRUE, show_output = TRUE)
```


```{r gen_code_blocks}

p_names <- paste0("p", seq_len(nrow(plane_pts)))
m_names <- paste0("m", 2:4)

block_list <- list()

for(i in seq_len(nrow(plane_pts))) {
  block_list[[i]] <- list()
  for(j in seq_len(ncol(plane_pts))) {
    ## this just shows the code that is getting executed
    rlang::expr(slim_block(fitness(!!slimr_name(m_names[j]), !!slimr_name(p_names[i])), {
      dominance = asInteger(homozygous) * 0.5 + 0.5;
      return(1 + (!!plane_pts[i, j] - sim.mutationFrequencies(subpop, mut) * dominance));
    })) %>%
      rlang::expr_print()
    
    ## actually make the slim_blocks
    block_list[[i]][[j]] <- slim_block(fitness(!!slimr_name(m_names[j]), !!slimr_name(p_names[i])), {
      dominance = asInteger(homozygous) * 0.5 + 0.5;
      return(1 + (!!plane_pts[i, j] - sim.mutationFrequencies(subpop, mut) * dominance));
    })
    
  }
}


```

