<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Simple Example of a Simulation using Nucleotides</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Simple Example of a Simulation using
Nucleotides</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="do">## set some parameters</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>seed <span class="ot">&lt;-</span> <span class="dv">1205</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>split_prob <span class="ot">&lt;-</span> <span class="fl">0.001</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>max_subpops <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="do">## specify simulation</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>split_isolate_sim <span class="ot">&lt;-</span> <span class="fu">slim_script</span>(</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>  </span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>  <span class="fu">slim_block</span>(<span class="fu">initialize</span>(), {</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>  </span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>    <span class="fu">setSeed</span>(<span class="sc">!!</span>seed);</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>    </span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a>    <span class="do">## tell SLiM to simulate nucleotides</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>    <span class="fu">initializeSLiMOptions</span>(<span class="at">nucleotideBased=</span>T);</span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>    <span class="fu">initializeAncestralNucleotides</span>(<span class="fu">randomNucleotides</span>(<span class="dv">1000</span>));</span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a>    <span class="fu">initializeMutationTypeNuc</span>(<span class="st">&quot;m1&quot;</span>, <span class="fl">0.5</span>, <span class="st">&quot;f&quot;</span>, <span class="fl">0.0</span>);</span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a>    </span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>    <span class="fu">initializeGenomicElementType</span>(<span class="st">&quot;g1&quot;</span>, m1, <span class="fl">1.0</span>, <span class="fu">mmJukesCantor</span>(<span class="fl">1e-5</span>));</span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a>    <span class="fu">initializeGenomicElement</span>(g1, <span class="dv">0</span>, <span class="dv">1000</span> <span class="sc">-</span> <span class="dv">1</span>);</span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a>    <span class="fu">initializeRecombinationRate</span>(<span class="fl">1e-8</span>);</span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a>    </span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a>  }),</span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a>  </span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a>  <span class="fu">slim_block</span>(<span class="dv">1</span>, {</span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a>    </span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a>    <span class="fu">defineGlobal</span>(<span class="st">&quot;curr_subpop&quot;</span>, <span class="dv">1</span>);</span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a>    <span class="fu">sim.addSubpop</span>(curr_subpop, <span class="dv">100</span>)</span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a>    </span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a>  }),</span>
<span id="cb1-30"><a href="#cb1-30" tabindex="-1"></a>  </span>
<span id="cb1-31"><a href="#cb1-31" tabindex="-1"></a>  <span class="fu">slim_block</span>(<span class="dv">1</span>, <span class="dv">10000</span>, <span class="fu">late</span>(), {</span>
<span id="cb1-32"><a href="#cb1-32" tabindex="-1"></a>    </span>
<span id="cb1-33"><a href="#cb1-33" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="sc">!!</span>split_prob) <span class="sc">==</span> <span class="dv">1</span>) {</span>
<span id="cb1-34"><a href="#cb1-34" tabindex="-1"></a>      <span class="do">## split a subpop</span></span>
<span id="cb1-35"><a href="#cb1-35" tabindex="-1"></a>      subpop_choose <span class="ot">=</span> <span class="fu">sample</span>(sim.subpopulations, <span class="dv">1</span>)</span>
<span id="cb1-36"><a href="#cb1-36" tabindex="-1"></a>      curr_subpop <span class="ot">=</span> curr_subpop <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb1-37"><a href="#cb1-37" tabindex="-1"></a>      <span class="fu">sim.addSubpopSplit</span>(<span class="at">subpopID =</span> curr_subpop, </span>
<span id="cb1-38"><a href="#cb1-38" tabindex="-1"></a>                         <span class="at">size =</span> <span class="dv">100</span>, </span>
<span id="cb1-39"><a href="#cb1-39" tabindex="-1"></a>                         <span class="at">sourceSubpop =</span> subpop_choose)</span>
<span id="cb1-40"><a href="#cb1-40" tabindex="-1"></a>      <span class="do">## if too many subpops, remove one randomly</span></span>
<span id="cb1-41"><a href="#cb1-41" tabindex="-1"></a>      <span class="cf">if</span>(<span class="fu">size</span>(sim.subpopulations) <span class="sc">&gt;</span> <span class="sc">!!</span>max_subpops) {</span>
<span id="cb1-42"><a href="#cb1-42" tabindex="-1"></a>        subpop_del <span class="ot">=</span> <span class="fu">sample</span>(sim.subpopulations, <span class="dv">1</span>)</span>
<span id="cb1-43"><a href="#cb1-43" tabindex="-1"></a>        <span class="fu">subpop_del.setSubpopulationSize</span>(<span class="dv">0</span>)</span>
<span id="cb1-44"><a href="#cb1-44" tabindex="-1"></a>      }</span>
<span id="cb1-45"><a href="#cb1-45" tabindex="-1"></a>    }</span>
<span id="cb1-46"><a href="#cb1-46" tabindex="-1"></a>    </span>
<span id="cb1-47"><a href="#cb1-47" tabindex="-1"></a>    <span class="fu">slimr_output_nucleotides</span>(<span class="at">subpops =</span> <span class="cn">TRUE</span>, <span class="at">do_every =</span> <span class="dv">100</span>)</span>
<span id="cb1-48"><a href="#cb1-48" tabindex="-1"></a>      </span>
<span id="cb1-49"><a href="#cb1-49" tabindex="-1"></a>  }),</span>
<span id="cb1-50"><a href="#cb1-50" tabindex="-1"></a>  </span>
<span id="cb1-51"><a href="#cb1-51" tabindex="-1"></a>  <span class="fu">slim_block</span>(<span class="dv">10000</span>, <span class="fu">late</span>(), {</span>
<span id="cb1-52"><a href="#cb1-52" tabindex="-1"></a>    <span class="fu">sim.simulationFinished</span>()</span>
<span id="cb1-53"><a href="#cb1-53" tabindex="-1"></a>  })</span>
<span id="cb1-54"><a href="#cb1-54" tabindex="-1"></a>  </span>
<span id="cb1-55"><a href="#cb1-55" tabindex="-1"></a>)</span>
<span id="cb1-56"><a href="#cb1-56" tabindex="-1"></a></span>
<span id="cb1-57"><a href="#cb1-57" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">slim_run</span>(split_isolate_sim)</span></code></pre></div>
<pre><code>## 
## 
## Simulation finished with exit status: 0
## 
## Success!</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>res_data <span class="ot">&lt;-</span> <span class="fu">slim_results_to_data</span>(results)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>res_data</span></code></pre></div>
<pre><code>## # A tibble: 100 × 6
##    type             expression                 genera…¹ name  data       subpops
##    &lt;chr&gt;            &lt;chr&gt;                         &lt;int&gt; &lt;chr&gt; &lt;list&gt;     &lt;list&gt; 
##  1 slim_nucleotides slimr_output_nucleotides()      100 seqs  &lt;DNAStrnS&gt; &lt;chr&gt;  
##  2 slim_nucleotides slimr_output_nucleotides()      200 seqs  &lt;DNAStrnS&gt; &lt;chr&gt;  
##  3 slim_nucleotides slimr_output_nucleotides()      300 seqs  &lt;DNAStrnS&gt; &lt;chr&gt;  
##  4 slim_nucleotides slimr_output_nucleotides()      400 seqs  &lt;DNAStrnS&gt; &lt;chr&gt;  
##  5 slim_nucleotides slimr_output_nucleotides()      500 seqs  &lt;DNAStrnS&gt; &lt;chr&gt;  
##  6 slim_nucleotides slimr_output_nucleotides()      600 seqs  &lt;DNAStrnS&gt; &lt;chr&gt;  
##  7 slim_nucleotides slimr_output_nucleotides()      700 seqs  &lt;DNAStrnS&gt; &lt;chr&gt;  
##  8 slim_nucleotides slimr_output_nucleotides()      800 seqs  &lt;DNAStrnS&gt; &lt;chr&gt;  
##  9 slim_nucleotides slimr_output_nucleotides()      900 seqs  &lt;DNAStrnS&gt; &lt;chr&gt;  
## 10 slim_nucleotides slimr_output_nucleotides()     1000 seqs  &lt;DNAStrnS&gt; &lt;chr&gt;  
## # … with 90 more rows, and abbreviated variable name ¹​generation</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="do">## sequences at generation 10000</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">image</span>(ape<span class="sc">::</span><span class="fu">as.DNAbin</span>(res_data<span class="sc">$</span>data[[<span class="dv">100</span>]]))</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAWlBMVEUAAAAAADoAAGYAAP8AOpAAZrYA/wA6AAA6ADo6AGY6kNtmAABmZjpmtv+QOgCQ2/+2ZgC225C2/7a2///bkDrb////AAD/tmb/trb/25D//wD//7b//9v///9So+0IAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAIDklEQVR4nO3cgXqiOBQFYNpM09m6dbo7TFijff/X3NybBKEqBwUp4DnzzYzaEMhviKZBi0+mM8V3H8DcQyAQAoEQCIRAIAQCIRAIgUAIBEIgEAKBEAiEQCAEAiEQCIFACARCIBACgRAIhEAgBAIhEAiBQAgEQiAQAoEQCIRAIAQCIRAIgUAIBEIgEAKBEAiEQCAEAiEQCIFACARCIBACgRAIhEAgBAIhEAiBQAgEQiCQ+wAVrcyzYM/cCcg10tkc30hnQdtIZ0HTyPKAPoq3dsELQPtNePp//GkUvAB02H4teAZIa5M8/76+KVdv0avWC0C7n3+3j/ECUFm8S9sbZS8AVQr+8fTrWPBCD9r9PJa5qik3bQVrvQBUPv/z+t4qeBZoFwsdti/HgmeB6oLHPrRooNDiZqM/LwF9pPaWR8zzQGUq+F+jqy0ZqAqnQvnUPNCzQIdte6DSgueAvmjHgksGkp6x3zRbfxZov2mdhrHgOaB2XanggoF2rzqiNofpDqCP1qvOQwCV6cW20T8uAKV2VwBoZadYas5+02hV9yCNgOpBujqiLxioSsNzc5jufJnHQOt6mc/9Ig5FqeClN4o6WhXofVB8o3jYLu+N4pkZ465+i9gYpi9NLXVy0BrNuwq2phoXCs4LaEUhEAiBQAgEQiAQAoEQCIRAIAQCIRAIgUAIBEIgkJGBvLXWOWuMD3+N9zbcDrec3I0Ph8eMD7edld9w2PAjEzYIBcIfY72TGuQR77WOWJVPP5AioRqt1IVSelcfdFK/VullYxv35nX7UM7Lj12872SD3u0eHcgTqBuIPYhABLorEMcgAg0D4ilGIAJNB2RHBXIrAArHGFCySDggpwcvd53Ro/TSdrmfgMKPfQLSxiiQ9BUFktsKpI2VTeQhpy8FGcjLPkLdNgIJmamBZLcJSHYUalOgzgtGCUQgAhGIQAQiEIEIRCACzQ0o/T7IJ6B08Ho3TdGkTTJdMvHCTeGUBuS5mLRSEE0Ekge1cTLxVVSRM7oXSWi9uIYtXJ6L6TQvzrgystDobpzMxeQZIBAA+u7fKPrYqfXEAkBmJKA8mzdLAZp3D5rBKdbRgzx70PinmF8dEE8xAg0D4ilGIAItCWg1L/NylHGQlimqHFCcu8rxmTxZlabp/FM9rasnq9p4k9oqN2VmGtsoys7GlTaXatX/XTTPk9WoGSerLk1Wvd5xcV3Myqx5rkAeA9mHAbK39aDHAWIPugGIYxCBCHRXII5B7EHDgNiDCISA0lxM16yMsJi8TGbTXEyvmZOmyVxJCuVL8OJKVpyf5XWxdAledBUcvepR1sOk5frrJOvqS/B0c9e6BE+EZIFOVsv0CXM6S+v/BYJzAvI9gOz1QJ5ABCIQxyD2IAIR6NuBOAYRiEB9gOIHeWxcZnE6ndJJkZNZkALpB4KsSe1rzcXiJXh6CYgu2cTL93Ti5nSyJUA+z8Xih3/0Mzz1JXhO65LVpDgXkyWfuOyUln0WDOT6AbmHBWIPYg9iDyIQgQhEoNUC8VVsHKBl96C0ECMSCUh/32BNvsLsdC5m67lYBrIZqDEX04vLZC4Wt9b2x9+A6FzM2QxkWnMx/bRPnvDFuZj/1mWfcYDOTFYRkOkNZAlEoIcH4hjEHjTjHrQGIPYgAvEUuyvQinuQjys0XhT0+GSupEC2BnItIOe+Arl62Sddr5+BXAJybSCrQPkDdfIE6Wda4rKPLD6plH9IIKdz1xMgSyACEYhABCLQ/IH4Mk8gnmIE+gL0ddknXWHWmIvJeaGzLJ+WfXQ9qF449I1lH99e9rEXln3cuWUfnz/MkpadGss+8wZy9bpYBrKdQOZGILNUIPagQT3INYBsZw/y7EHSgwx7EHvQFIO0JRBfxTgGEYiDdB8gaa9qSG/QuZhvzsVkNqWfodfEb9vyPgNZVdCfx66jX9Fl41d0xYWfuOyjsAok+4iLQWlK59JByBTM69b6xc0JyLkZfIfZCEB+GJBZEFD6kjc/FpA5C2T6ApnZAX1fD7LsQavoQQSaGMitDWiaMagTyM56DOLLPADiGDQx0OrGIAI1q1tMvglojB2OUGKMnVxbcKwQaPAOCXT/EgSaYCfXFhwrBBq8QwLdvwSBJtjJtQUfNQQCIRAIgUAIBEIgEAKBEAiEQCAEAiEQCIFACAQyBdBhWxTFm9yqiuLpV+tG7+xei+JlSB27v36f37y7ngmADtuw/1JaV4Ub8vd4o3eqH38+95sBdew3zwJ0ujmoZwKg3et7+Ld8/n3YSjf6ePmsb/RO3GJAHaGbCNDp5qieycag8CTVUvWN3lvvfqan+MY6quKtkqKnm6N6JgP6CMeirawaN3pvXT3/u9Fx7PY6ItDJ5qieqYDCU5jO9PBvfaP35qWcIIfty4A61OB0c1TPREBVHqNvBXpKz/NKgSp9lR9wisUxIowX6zzFyvguaMAgHRsQGnN7HTMepMviXf8f8DK/30gV1ZC3CtVsX+Z3r2/5GG9/o1iGN4ralpvrqGb7RrGMl3TJMZT5XX159VSjytOVW+tIw8zp5t31cLIKQiAQAoEQCIRAIAQCIRAIgUAIBEIgEAKBEAiEQCAEAiEQCIFACARCIBACgRAIhEAgBAIhEAiBQAgEQiAQAoEQCIRAIAQCIRAIgUAIBEIgEAKBEAiEQCAEAiEQCIFACARCIBACgRAIhEAgBAIhEAiBQAgEQiAQAoEQCIRAIAQCIRAIgUAIBEIgEAKBEAiEQCAEAiEQyP/VKNqEvHOI7AAAAABJRU5ErkJggg==" /><!-- --></p>
<p>And then we can use some other R packages to quickly build a tree
based on the simulated nucleotides, to see if it looks like what we
would expect from a sequentially splitting population.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="do">## convert to ape::DNAbin</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>al <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">as.DNAbin</span>(res_data<span class="sc">$</span>data[[<span class="dv">100</span>]])</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>dists <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">dist.dna</span>(al)</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>upgma_tree <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">as.phylo</span>(<span class="fu">hclust</span>(dists, <span class="at">method =</span> <span class="st">&quot;average&quot;</span>))</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>pal <span class="ot">&lt;-</span> paletteer<span class="sc">::</span><span class="fu">paletteer_d</span>(<span class="st">&quot;RColorBrewer::Paired&quot;</span>, <span class="dv">10</span>)</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="fu">plot</span>(upgma_tree, <span class="at">show.tip.label =</span> <span class="cn">FALSE</span>)</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>ape<span class="sc">::</span><span class="fu">tiplabels</span>(<span class="at">pch =</span> <span class="dv">19</span>, <span class="at">col =</span> pal[<span class="fu">as.numeric</span>(<span class="fu">as.factor</span>(res_data<span class="sc">$</span>subpops[[<span class="dv">100</span>]]))])</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAJFBMVEUAAAAfeLQzoCxqPZqmzuOy34rKstbjGhz7mpn9v2//fwD///+btxhlAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAD0ElEQVR4nO3YWXKjMAAAUXnB6/3vGwsylZqUUVtBgEDdv9aWV0Cww9OShbUPUHsCQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBzQF1XZc1vjWgrssUagyo63KFBIIEggSCGgIKrwRKFP9UgRIJBDUJFHJ6tgiUOVYgGCtQauwO/81/9FzJaXdAkwf8SiDo0PscSh5h1QoDhXCIQgeBRkcfhkoeYdUYKK/2gHKXEwiW2whQ1tenovtuBaj4wE8TCBIIEgi69D6XjBltAYVLFLoINLreZShnSuEjlN0080WZ2xtQ8Y1rAlru/S/jTFUBrbEpJBAlECQQJBB07n3OGTNaA4pCZ4He7/rqPJQza77zrLFpqrirQIm2BLROz80ArdMAdLvdBHpfD3SLrQ607jfSseKZbkM5s2Y5ymwrT00gSKB03mLJfAaN9PMiJNDb/p1k/1fQhFfo7/YONHmaQG+n/SQQJFB6fl1X0AxNPldNQBWuG5733ue+4J7LLVpg3RdQFLoLNDY/3IcW3HO5RQusG6+gfQNNfsjvHGhqEeh4PAo01gvoGBNopBCOQzlzZjnIHIsWSSBIoHTeYpBAo32/CAk01nCkxoE++KWkcSD+TKD0Z63fYqnPfEh/su1GgSb/ivH/gzjVRoGW2jaEU+9zKrnoX6oW6HmKQieBRgechkou+pfWAuKHVNtAH2wkEGwkUHqjUB1QqfeXUtUHtNoZ3icQJBAkEPTofR4ZM1oDikIPgUZ7DGXMqAJoubYJtGACQbUArfGS/FGVAC22fHYCQQJBAkHX3ueaMaM1oCh0FWi061DGDIEggaCWgF7vrQKlCjVeQVVVIVBVhQpvsdoSiMr1aQ7omenTHlBuAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQV+5PKNvNjGhHgAAAABJRU5ErkJggg==" /><!-- --></p>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
